### ISSUE TITLE
[enhancement] Install the colormap of the focused window

### ISSUE THREAD

**Issue by Airblader** [i3/i3#2451]
We currently don't use `xcb_install_colormap` at all. From what I understand we should ensure to install the »correct« colormap. Sources disagree on the number of colormaps that can be installed in parallel, but at least it seems like usually the limit is 1. Therefore it makes sense to have the colormap of the focused window be the installed one.

Openbox does exactly this while awesome seems to ignore it like i3 currently does (@psychon Are you aware of this? Is there a specific decision behind this in awesome or has it just never come up?)

I wouldn't say we should implement this right away, but I want to open this ticket to track it in case problems arise. That way we also have somebody to test the change.

**Comment by psychon** [comment]
The 90's called. They want their issues back. More honestly: I wouldn't even know how to test this.

With 256 colors, the colormap defines what color 0/1/2/3 etc actually is on screen. I think the technical detail is the visual's class (e.g. XCB_VISUAL_CLASS_STATIC_COLOR etc). According to `xdpyinfo`, my X server currently only offers TrueColor and DirectColor visuals. The root window's default visual (which is what everyone uses) has class TrueColor and the depth=32 visual (which is what everyone uses who doesn't use the default visual) also has class TrueColor. Google points me to https://tronche.com/gui/x/xlib/window/visual-types.html for an explanation. This says that for TrueColor visuals the colormap is read-only, so you cannot set the colormap anyway.

TL;DR: It has never come up and I doubt a lot that it will ever come up.

**Comment by stapelberg** [comment]
The only use case I’m aware of for non-TrueColor visuals is the Raspberry Pi. Not sure if colormaps are an issue on the Raspberry Pi, though.

**Comment by Airblader** [comment]
I imagine they aren't since we did changes for this in 287a0b4c and the user seemed to be OK with our colormap handling then. Perhaps we should close this issue again, it may not be worth tracking.

**Comment by psychon** [comment]
Sigh. Stop linking commits. ;-)
287a0b4 is wrong. I left a comment on the commit and leave handling that to someone else.

For the Raspberry Pi: Are you sure it has a non-TrueColor visual? And you'd still need to find an app that doesn't just "live" with the default colormap, but wants its own and invokes the ICCCM mechanisms for asking the WM to set a colormap.

**Comment by Airblader** [comment]
Let's close this again.
